File: articles/education-schooling-and-developmental-formation/cpp/schooling_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) values.push_back(item);
    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) return static_cast<int>(i);
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: schooling_summary data/schooling_development_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);
    auto header = split_csv_line(line);

    int development_col = find_column(header, "development_score");
    int connectedness_col = find_column(header, "connectedness_score");
    int teacher_col = find_column(header, "current_teacher");
    int peer_col = find_column(header, "current_peer");
    int stress_col = find_column(header, "current_stress");

    long count = 0;
    double development_sum = 0.0, connectedness_sum = 0.0, teacher_sum = 0.0, peer_sum = 0.0, stress_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);
        if ((int)values.size() <= development_col || (int)values.size() <= connectedness_col || (int)values.size() <= teacher_col || (int)values.size() <= peer_col || (int)values.size() <= stress_col) continue;
        development_sum += std::stod(values[development_col]);
        connectedness_sum += std::stod(values[connectedness_col]);
        teacher_sum += std::stod(values[teacher_col]);
        peer_sum += std::stod(values[peer_col]);
        stress_sum += std::stod(values[stress_col]);
        count++;
    }

    std::cout << "Rows analyzed: " << count << "\n";
    std::cout << "Mean development_score: " << development_sum / count << "\n";
    std::cout << "Mean connectedness_score: " << connectedness_sum / count << "\n";
    std::cout << "Mean current_teacher: " << teacher_sum / count << "\n";
    std::cout << "Mean current_peer: " << peer_sum / count << "\n";
    std::cout << "Mean current_stress: " << stress_sum / count << "\n";
    return 0;
}

File: articles/education-schooling-and-developmental-formation/go/schooling_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run schooling_summary.go data/schooling_development_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	developmentIdx := col(header, "development_score")
	connectednessIdx := col(header, "connectedness_score")
	teacherIdx := col(header, "current_teacher")
	peerIdx := col(header, "current_peer")
	stressIdx := col(header, "current_stress")
	if developmentIdx < 0 || connectednessIdx < 0 || teacherIdx < 0 || peerIdx < 0 || stressIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	developmentSum, connectednessSum, teacherSum, peerSum, stressSum := 0.0, 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		development, e1 := strconv.ParseFloat(row[developmentIdx], 64)
		connectedness, e2 := strconv.ParseFloat(row[connectednessIdx], 64)
		teacher, e3 := strconv.ParseFloat(row[teacherIdx], 64)
		peer, e4 := strconv.ParseFloat(row[peerIdx], 64)
		stress, e5 := strconv.ParseFloat(row[stressIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil || e5 != nil {
			continue
		}
		count++
		developmentSum += development
		connectednessSum += connectedness
		teacherSum += teacher
		peerSum += peer
		stressSum += stress
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean development_score: %.4f\n", developmentSum/float64(count))
	fmt.Printf("Mean connectedness_score: %.4f\n", connectednessSum/float64(count))
	fmt.Printf("Mean current_teacher: %.4f\n", teacherSum/float64(count))
	fmt.Printf("Mean current_peer: %.4f\n", peerSum/float64(count))
	fmt.Printf("Mean current_stress: %.4f\n", stressSum/float64(count))
}
